bam_checksum.c

```diff
@@ -0,0 +1,1307 @@
+            else                   code = 'I';
+        } else {
+            if      (val >= -128   && val <= 127)   code = 'c';
+            else if (val >= -32768 && val <= 32767) code = 's';
+            else                                    code = 'i';
+        }
+        if (code == tag[2])
+            // Already optimal.  The usual code path
+            return tag;
+
+        // Otherwise rewrite it;
+        ct[0] = tag[0];
+        ct[1] = tag[1];
+        ct[2] = code;
+        switch (code) {
+        case 'C': case 'c':
+            ct[3] = val;
+            *tag_len = 4;
+            break;
+
+        case 'S': case 's':
+            // Don't care about sign as it's defined anyway
+            u16_to_le(val, ct+3);
+            *tag_len = 5;
+            break;
+
+        case 'I': case 'i':
+            // Don't care about sign as it's defined anyway
+            u32_to_le(val, ct+3);
+            *tag_len = 7;
+            break;
+        }
+        return ct;
+
+    default:
+        return tag;
+    }
+}
+
+// Qsort callback, by integer
+static int tag_qsort(const void *t1, const void *t2) {
+    return *(const int *)t1 - *(const int *)t2;
+}
+
+/*
+ * Produces a concatenated string of aux tags in <ID><TYPE><VAL> binary
+ * representation,  with the tag names and orders defined in tag_ids[],
+ * checksums it, and combines it with the flag-seq CRC.
+ * If *tag_str is "*" then we negate tag_ids and encode everything but those.
+ * This is a bit trickier as we can no longer use the order specified and
+ * instead encode in ASCII sorted order instead.
+ *
+ * If the read-group is found in the RG:Z: aux, this is returned in
+ * the *RGZ ptr (which points to the <VAL> field.
+ *
+ * Returns 0 on success, updating *crc_aux,
+ *        -1 on error
+ */
+int hash_aux(bam1_t *b, kstring_t *ks, int ntags,
+             char **tag_ids,
+             uint8_t **tag_ptr, size_t *tag_len,
+             const char *tag_str, short (*tag_keep)[75],
+             uint32_t crc_seq, uint32_t *crc_aux,
+             uint8_t **RGZ) {
+    size_t aux_len = bam_get_l_aux(b);
+    // 1 byte minimum forces a non-NULL pointer so CRC works
+    if (ks_resize(ks, aux_len+1) < 0)
+        return -1;
+    uint8_t *aux_ptr = (uint8_t *)ks->s;
+
+    // Pass 1: find all tags to copy and their lengths
+    uint8_t *aux = bam_aux_first(b), *aux_next;
+    memset(tag_len, 0, ntags * sizeof(*tag_len));
+    int tag_id[4000]; // a-zA-Z0-9 is 62. 62^2 is 3844
+
+    if (*tag_str == '*') {
+        // All tags bar specific ones, in alphanumeric order.
+        // Select the tags by name on pass 1, then sort by name to get
+        // a canonical order, and finally concatenate tags in order.
+        ntags = 0;
+        while (aux) {
+            if (aux[-2] == 'R' && aux[-1] == 'G' && aux[0] == 'Z' && RGZ)
+                *RGZ = aux+1;
+            aux_next = bam_aux_next(b, aux);
+            if (tag_keep[aux[-2]-'0'][aux[-1]-'0'] == 0) {
```

Comment:
I'm not sure why I chose to only have official printable tag name space here rather than just 256x256.  I assume it was to try and speed it up by keeping the lookup table small enough for cache.  I assume I tried it at some point, but cannot recall.  Anyway the simple bounds check has been added now.  Thanks.

Reply:
On looking again, I noticed that the check isn't quite right:
```
aux[-1] >= '0' && aux[-2] <= 'z'
```
should be
```
aux[-1] >= '0' && aux[-1] <= 'z'
```
(and the same below on line 487).

Reply:
Oops.  Indeed.  Good spot.